src/auth.tsx

```diff
@@ -59,9 +60,136 @@ export interface ClaimsCheckProps {
+ * const {status, data: signInCheckResult} = useSignInCheck({forceRefresh: true, requiredClaims: {admin: true}});
+ * ```
+ */
+export function useSigninCheck(
+  options?: SignInCheckOptionsBasic | SignInCheckOptionsClaimsObject | SignInCheckOptionsClaimsValidator
+): ObservableStatus<SigninCheckResult> {
+  // If both `requiredClaims` and `validateClaims` are provided, we won't know which one to use
+  if (options?.hasOwnProperty('requiredClaims') && options?.hasOwnProperty('validateClaims')) {
+    throw new Error('Cannot have both "requiredClaims" and "validateClaims". Use one or the other.');
+  }
+
+  const auth = useAuth();
+
+  // ObservableId should change for different options configurations to ensure no cache collisions
+  let observableId = `auth:signInCheck:${auth.app.name}::forceRefresh:${!!options?.forceRefresh}`;
+  if (options?.forceRefresh) {
+    observableId = `${observableId}:forceRefresh:${options.forceRefresh}`;
+  }
+  if (options?.hasOwnProperty('requiredClaims')) {
+    observableId = `${observableId}:requiredClaims:${JSON.stringify((options as SignInCheckOptionsClaimsObject).requiredClaims)}`;
+  } else if (options?.hasOwnProperty('validateCustomClaims')) {
+    observableId = `${observableId}:validateClaims:${JSON.stringify((options as SignInCheckOptionsClaimsValidator).validateCustomClaims)}`;
+  }
+
+  const observable = user(auth).pipe(
+    switchMap(user => {
+      if (!user) {
+        return of({ signedIn: false, hasRequiredClaims: false });
+      } else if (options && (options.hasOwnProperty('requiredClaims') || options.hasOwnProperty('validateClaims'))) {
+        return from(user.getIdTokenResult(options?.forceRefresh ?? false)).pipe(
+          map(idTokenResult => {
+            let validator: ClaimsValidator;
+
+            if (options.hasOwnProperty('requiredClaims')) {
+              validator = getClaimsObjectValidator((options as SignInCheckOptionsClaimsObject).requiredClaims);
+            } else {
+              validator = (options as SignInCheckOptionsClaimsValidator).validateCustomClaims;
+            }
+
+            const { hasRequiredClaims, errors } = validator(idTokenResult.claims);
+            return { signedIn: true, hasRequiredClaims, errors, user: user };
+          })
+        );
+      } else {
+        // If no claims are provided to be checked, `hasRequiredClaims` is true
+        return of({ signedIn: true, hasRequiredClaims: true, user: user });
+      }
+    })
+  );
+
+  return useObservable(observableId, observable);
+}
+
+function getClaimsObjectValidator(requiredClaims: firebase.auth.IdTokenResult['claims']): ClaimsValidator {
+  return function claimsObjectValidator(userClaims) {
+    const errors: { [key: string]: String[] } = {};
+
+    Object.keys(requiredClaims).forEach(claim => {
+      if (requiredClaims[claim] !== userClaims[claim]) {
+        errors[claim] = [`Expected "${requiredClaims[claim]}", but user has "${userClaims[claim]}" instead`];
```

Comment:
change to `new ClaimsCheckError`

Reply:
https://github.com/firebase/firebase-js-sdk/blob/5ad7ff2ae955c297556223e6cb3ad9d4b897f664/packages/util/src/errors.ts#L74

Reply:
or FirebaseError?

Reply:
made ReactFireError